iOS enabled mock and low level tests

yaronyg opened the pull request:
This checkin switches us to native mode testing for Android.

The checkin contains the code to teach the wifiBasedNativeMock how to 'speak' iOS and updates all the testThaliMobileNative*.js tests to be compliant with the new iOS APIs and passes with the mock on all those tests.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/thaliproject/thali_cordovaplugin/1275)
<!-- Reviewable:end -->

ThaliTester commented:
PR is added to the queue for testing as 1. task. (915c2f2)

ThaliTester commented:
Test (Success) 88496963 build is completed (915c2f2)

See https://github.com/ThaliTester/TestResults/tree/88496963915c2f2_iOS_enabled_mock_and_low_level_tests_yaronyg/ for the logs

ThaliTester commented:
Test 88496963915c2f2(915c2f2) has failed

See https://github.com/ThaliTester/TestResults/tree/88496963915c2f2_iOS_enabled_mock_and_low_level_tests_yaronyg/ for the fail logs

chapko commented:
@yaronyg there is small change that makes one test fail, but what is more interesting is that why it does not fail on CI during build, it didn't even try to run this test (testThaliMobileNativeDiscoveryCoordinated.js)

---

Reviewed 11 of 11 files at r2.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[test/www/jxcore/bv_tests/testThaliMobileNative.js, line 319 at r2](https://reviewable.io:443/reviews/thaliproject/thali_cordovaplugin/1275#-KTiEMWlgyIgx3MtHqLK:-KTiEMWmveupSNhhBCK5:bap694z) ([raw file](https://github.com/thaliproject/thali_cordovaplugin/blob/915c2f2e94c4a4ca06906786409c9947476c9976/test/www/jxcore/bv_tests/testThaliMobileNative.js#L319)):*
> ```JavaScript
>           }, 3000);
>         });
>       });
> ```

NB: I think we should agree on one indentation styling, because it looks very confusing for me. Test starts at 0 indentation level and ends at the 3rd level:
```javascript
test('test name ...',
  function () {
    //...
      });
```

---

*[test/www/jxcore/bv_tests/testThaliMobileNativeDiscoveryCoordinated.js, line 42 at r2](https://reviewable.io:443/reviews/thaliproject/thali_cordovaplugin/1275#-KTiPFSjQxMUZT_cfehf:-KTiPFSjQxMUZT_cfehg:b-grr5c3) ([raw file](https://github.com/thaliproject/thali_cordovaplugin/blob/915c2f2e94c4a4ca06906786409c9947476c9976/test/www/jxcore/bv_tests/testThaliMobileNativeDiscoveryCoordinated.js#L42)):*
> ```JavaScript
>     }
> 
>     return Promise.resolve(null);
> ```

NB: this line is identical to `return null;`. Also next `.then` in chain does not use this value anyway so I don't see the point of explicit return.

---

*[test/www/jxcore/bv_tests/testThaliMobileNativeDiscoveryCoordinated.js, line 127 at r2](https://reviewable.io:443/reviews/thaliproject/thali_cordovaplugin/1275#-KTiTvODfrfUXOJcaXMh:-KTiTvODfrfUXOJcaXMi:b-9nh6x8) ([raw file](https://github.com/thaliproject/thali_cordovaplugin/blob/915c2f2e94c4a4ca06906786409c9947476c9976/test/www/jxcore/bv_tests/testThaliMobileNativeDiscoveryCoordinated.js#L127)):*
> ```JavaScript
> 
>   var currentPeers = {};
>   latestPeers = {};
> ```

this change changes test logic.
Previously the following was true:
```
currentPeers === latestPeers
```
Now it is not.

'check latest peer discovery' fails on desktop, and I have no idea why all these tests are not running on CI during build.

---

*[test/www/jxcore/bv_tests/testThaliMobileNativeiOS.js, line 6 at r2](https://reviewable.io:443/reviews/thaliproject/thali_cordovaplugin/1275#-KTiYZ6jjXd4tUoEWA6n:-KTiYZ6kAre1APhdU-Wf:b-1fx4jc) ([raw file](https://github.com/thaliproject/thali_cordovaplugin/blob/915c2f2e94c4a4ca06906786409c9947476c9976/test/www/jxcore/bv_tests/testThaliMobileNativeiOS.js#L6)):*
> ```JavaScript
> var ThaliMobile = require('thali/NextGeneration/thaliMobile');
> var Platform = require('thali/NextGeneration/utils/platform');
> var nodeUuid = require('node-uuid');
> ```

NB: it used to be `uuid` without `node` prefix.

---

*[test/www/jxcore/lib/testUtils.js, line 3 at r2](https://reviewable.io:443/reviews/thaliproject/thali_cordovaplugin/1275#-KTie-I3ISwEwyFSdgL1:-KTie-I4F09moFXg60x_:b-iw49xf) ([raw file](https://github.com/thaliproject/thali_cordovaplugin/blob/915c2f2e94c4a4ca06906786409c9947476c9976/test/www/jxcore/lib/testUtils.js#L3)):*
> ```JavaScript
> 'use strict';
> 
> var util = require('util');
> ```

NB: `util` is not used too.

---

*[test/www/jxcore/lib/thaliMobileNativeTestUtils.js, line 3 at r2](https://reviewable.io:443/reviews/thaliproject/thali_cordovaplugin/1275#-KTivL5uCQcBGG1Y7Vp8:-KTivL5uCQcBGG1Y7Vp9:b-as9uw6) ([raw file](https://github.com/thaliproject/thali_cordovaplugin/blob/915c2f2e94c4a4ca06906786409c9947476c9976/test/www/jxcore/lib/thaliMobileNativeTestUtils.js#L3)):*
> ```JavaScript
> 'use strict';
> 
> var Platform = require('thali/NextGeneration/utils/platform');
> ```

NB: platform should start with lowercase.

---

*[test/www/jxcore/lib/thaliMobileNativeTestUtils.js, line 120 at r2](https://reviewable.io:443/reviews/thaliproject/thali_cordovaplugin/1275#-KTivd3bqOfHpDsI0oaW:-KTivd3bqOfHpDsI0oaX:bnz3m9q) ([raw file](https://github.com/thaliproject/thali_cordovaplugin/blob/915c2f2e94c4a4ca06906786409c9947476c9976/test/www/jxcore/lib/thaliMobileNativeTestUtils.js#L120)):*
> ```JavaScript
> 
> module.exports.multiConnectEmitter =
>   new multiConnectEmitter();
> ```

it would be simpler without inheritance. Just
```javascript
module.exports.multiConnectEmitter = new EventEmitter();
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/thaliproject/thali_cordovaplugin/1275#-:-KTjIKofxcU3PclUUlb6:bu6n1fi)*
<!-- Sent from Reviewable.io -->

ThaliTester commented:
PR is added to the queue for testing as 1. task. (2a6a9b7)

yaronyg commented:
Remember this is the iOS branch so until #899 is done running tests on CI is going to basically be impossible. But I did run:

jx runCoordinatedTests.js --filter bv_tests/testMakeIntoCloseAllServer.js bv_tests/testPromiseQueue.js bv_tests/testTests.js bv_tests/testThaliMobileNative.js bv_tests/testThaliMobileNativeAndroid.js bv_tests/testThaliMobileNativeDiscoveryCoordinated.js bv_tests/testThaliMobileNativeiOS.js bv_tests/testThaliWifiInfrastructure.js bv_tests/testUsn.js 

Which I think are more or less all the tests that can run without #899, in NATIVE mode set to Android and they passed.

---

Review status: 7 of 11 files reviewed at latest revision, 9 unresolved discussions.

---

*[test/www/jxcore/bv_tests/testThaliMobileNative.js, line 319 at r2](https://reviewable.io:443/reviews/thaliproject/thali_cordovaplugin/1275#-KTiEMWlgyIgx3MtHqLK:-KTk4warnurPmIg_k2cl:br7065s) ([raw file](https://github.com/thaliproject/thali_cordovaplugin/blob/915c2f2e94c4a4ca06906786409c9947476c9976/test/www/jxcore/bv_tests/testThaliMobileNative.js#L319)):*
<details><summary><i>Previously, chapko (Eugene Chapko) wrote…</i></summary>
> NB: I think we should agree on one indentation styling, because it looks very confusing for me. Test starts at 0 indentation level and ends at the 3rd level:
> ```javascript
> test('test name ...',
>   function () {
>     //...
>       });
> ```
</details>
So my personal rule on things like this is that if it can't be enforced by a tool then I don't care. But I ran re-indent so hopefully that will be prettier.

---

*[test/www/jxcore/bv_tests/testThaliMobileNativeDiscoveryCoordinated.js, line 42 at r2](https://reviewable.io:443/reviews/thaliproject/thali_cordovaplugin/1275#-KTiPFSjQxMUZT_cfehf:-KTk6KAtw9nmwPX7o_K4:bki1n99) ([raw file](https://github.com/thaliproject/thali_cordovaplugin/blob/915c2f2e94c4a4ca06906786409c9947476c9976/test/www/jxcore/bv_tests/testThaliMobileNativeDiscoveryCoordinated.js#L42)):*
<details><summary><i>Previously, chapko (Eugene Chapko) wrote…</i></summary>
> NB: this line is identical to `return null;`. Also next `.then` in chain does not use this value anyway so I don't see the point of explicit return.
</details>
The point of an explicit return is to make Bluebird happy. It has very aggressive warning logic that throws a warning any time it sees a promise return undefined. They want explicit response values as a best practice. So this and the next Promise.resolve(null)s were an attempt to make Bluebird happy. But you are right, a simple null would do nicely.

---

*[test/www/jxcore/bv_tests/testThaliMobileNativeDiscoveryCoordinated.js, line 127 at r2](https://reviewable.io:443/reviews/thaliproject/thali_cordovaplugin/1275#-KTiTvODfrfUXOJcaXMh:-KTkj8MnjF9ME6x48cDv:bcmn7b8) ([raw file](https://github.com/thaliproject/thali_cordovaplugin/blob/915c2f2e94c4a4ca06906786409c9947476c9976/test/www/jxcore/bv_tests/testThaliMobileNativeDiscoveryCoordinated.js#L127)):*
<details><summary><i>Previously, chapko (Eugene Chapko) wrote…</i></summary>
> this change changes test logic.
> Previously the following was true:
> ```
> currentPeers === latestPeers
> ```
> Now it is not.
> 
> 'check latest peer discovery' fails on desktop, and I have no idea why all these tests are not running on CI during build.
</details>
So this line was previously:
 var currentPeers = latestPeers = {};

Now pick your favorite Javascript repl and run:
var latestPeers = { foo: 44 };
var currentPeers = latestPeers = {};
console.log('currentPeers:' + JSON.stringify(currentPeers));
console.log('latestPeers:' + JSON.stringify(latestPeers));

You will see that currentPeers and latestPeers are both set to {}

This is because the assignment operator in Javascript moves from the right to the left. See https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Assignment_Operators#Assignment

Which means that
var currentPeers = latestPeers = {} is exactly identical to what I have above.

BTW, the problem with the test was that latestPeers wasn't being set correctly and the previous line I replaced was also wrong so this test should have always failed.

---

*[test/www/jxcore/bv_tests/testThaliMobileNativeDiscoveryCoordinated.js, line 212 at r3](https://reviewable.io:443/reviews/thaliproject/thali_cordovaplugin/1275#-KTl4ELaQENAoh1j-2TS:-KTl4ELaQENAoh1j-2TT:b-w3nl1e) ([raw file](https://github.com/thaliproject/thali_cordovaplugin/blob/2a6a9b75451b6f75ba9f563128d397a0dc1cc083/test/www/jxcore/bv_tests/testThaliMobileNativeDiscoveryCoordinated.js#L212)):*
> ```JavaScript
>       peersReceived.forEach(function (idAndGeneration) {
>         allPeers[idAndGeneration] = true;
>         latestPeers[idAndGeneration] = true;
> ```

I am conflicted here. I could have just gotten rid of latestPeers and used allPeers but by using latestPeers I make sure that we are in fact just checking the latestPeers so if we have some ancient peer who really should be gone, if it shows up, we'll catch that error because we are only looking for the latest.

---

*[test/www/jxcore/bv_tests/testThaliMobileNativeiOS.js, line 6 at r2](https://reviewable.io:443/reviews/thaliproject/thali_cordovaplugin/1275#-KTiYZ6jjXd4tUoEWA6n:-KTkxg58NG4RIrjAqlJ_:b-n0g1ie) ([raw file](https://github.com/thaliproject/thali_cordovaplugin/blob/915c2f2e94c4a4ca06906786409c9947476c9976/test/www/jxcore/bv_tests/testThaliMobileNativeiOS.js#L6)):*
<details><summary><i>Previously, chapko (Eugene Chapko) wrote…</i></summary>
> NB: it used to be `uuid` without `node` prefix.
</details>
That is intentional. The variable name should be the same as the dependency.

---

*[test/www/jxcore/lib/testUtils.js, line 3 at r2](https://reviewable.io:443/reviews/thaliproject/thali_cordovaplugin/1275#-KTie-I3ISwEwyFSdgL1:-KTkxwkxGxN5SxmnU9wM:ba3dfob) ([raw file](https://github.com/thaliproject/thali_cordovaplugin/blob/915c2f2e94c4a4ca06906786409c9947476c9976/test/www/jxcore/lib/testUtils.js#L3)):*
<details><summary><i>Previously, chapko (Eugene Chapko) wrote…</i></summary>
> NB: `util` is not used too.
</details>
Removed

---

*[test/www/jxcore/lib/thaliMobileNativeTestUtils.js, line 3 at r2](https://reviewable.io:443/reviews/thaliproject/thali_cordovaplugin/1275#-KTivL5uCQcBGG1Y7Vp8:-KTky8ZnwtGUzbQ2qKkF:b-886gsx) ([raw file](https://github.com/thaliproject/thali_cordovaplugin/blob/915c2f2e94c4a4ca06906786409c9947476c9976/test/www/jxcore/lib/thaliMobileNativeTestUtils.js#L3)):*
<details><summary><i>Previously, chapko (Eugene Chapko) wrote…</i></summary>
> NB: platform should start with lowercase.
</details>
Fixed

---

*[test/www/jxcore/lib/thaliMobileNativeTestUtils.js, line 120 at r2](https://reviewable.io:443/reviews/thaliproject/thali_cordovaplugin/1275#-KTivd3bqOfHpDsI0oaW:-KTkyE9zMzceiBi4rrE6:b-886gsx) ([raw file](https://github.com/thaliproject/thali_cordovaplugin/blob/915c2f2e94c4a4ca06906786409c9947476c9976/test/www/jxcore/lib/thaliMobileNativeTestUtils.js#L120)):*
<details><summary><i>Previously, chapko (Eugene Chapko) wrote…</i></summary>
> it would be simpler without inheritance. Just
> ```javascript
> module.exports.multiConnectEmitter = new EventEmitter();
> ```
</details>
Fixed

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/thaliproject/thali_cordovaplugin/1275#-:-KTl1iFLygkty1gs3XUQ:b-qojl8i)*
<!-- Sent from Reviewable.io -->

ThaliTester commented:
Test (Success) 88496963 build is completed (2a6a9b7)

See https://github.com/ThaliTester/TestResults/tree/884969632a6a9b7_iOS_enabled_mock_and_low_level_tests_yaronyg/ for the logs

ThaliTester commented:
Test 884969632a6a9b7(2a6a9b7) has failed

See https://github.com/ThaliTester/TestResults/tree/884969632a6a9b7_iOS_enabled_mock_and_low_level_tests_yaronyg/ for the fail logs

chapko commented:
@yaronyg I added comments to the `latestPeers` issue, it is non-blocking so you can either remove this variable or leave it as is. The only thing left is `multiConnectEmitter` which need to be fixed and then it <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

---

Reviewed 4 of 4 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[test/www/jxcore/bv_tests/testThaliMobileNativeDiscoveryCoordinated.js, line 127 at r2](https://reviewable.io:443/reviews/thaliproject/thali_cordovaplugin/1275#-KTiTvODfrfUXOJcaXMh:-KTnGEO2GlTICWsxmzQM:b-4h93f0) ([raw file](https://github.com/thaliproject/thali_cordovaplugin/blob/915c2f2e94c4a4ca06906786409c9947476c9976/test/www/jxcore/bv_tests/testThaliMobileNativeDiscoveryCoordinated.js#L127)):*
<details><summary><i>Previously, yaronyg (Yaron Y Goland) wrote…</i></summary>
> So this line was previously:
>  var currentPeers = latestPeers = {};
> 
> Now pick your favorite Javascript repl and run:
> var latestPeers = { foo: 44 };
> var currentPeers = latestPeers = {};
> console.log('currentPeers:' + JSON.stringify(currentPeers));
> console.log('latestPeers:' + JSON.stringify(latestPeers));
> 
> You will see that currentPeers and latestPeers are both set to {}
> 
> This is because the assignment operator in Javascript moves from the right to the left. See https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Assignment_Operators#Assignment
> 
> Which means that
> var currentPeers = latestPeers = {} is exactly identical to what I have above.
> 
> BTW, the problem with the test was that latestPeers wasn't being set correctly and the previous line I replaced was also wrong so this test should have always failed.
</details>
I'm still not agree. The following will be the same as an old code:
```javascript
var obj = {};
var currentPeers = obj;
latestPeers = obj;
```
In your case you create 2 separate independent object. Initially they are both empty and look the same when serialized, but they do not affect each other anymore.

But it doesn't matter anymore, you have fixed failing test below by explicitly populating `latestPeers`.

---

*[test/www/jxcore/bv_tests/testThaliMobileNativeDiscoveryCoordinated.js, line 212 at r3](https://reviewable.io:443/reviews/thaliproject/thali_cordovaplugin/1275#-KTl4ELaQENAoh1j-2TS:-KTnKRKEvkaJgUpv0cfj:b2oj5gh) ([raw file](https://github.com/thaliproject/thali_cordovaplugin/blob/2a6a9b75451b6f75ba9f563128d397a0dc1cc083/test/www/jxcore/bv_tests/testThaliMobileNativeDiscoveryCoordinated.js#L212)):*
<details><summary><i>Previously, yaronyg (Yaron Y Goland) wrote…</i></summary>
> I am conflicted here. I could have just gotten rid of latestPeers and used allPeers but by using latestPeers I make sure that we are in fact just checking the latestPeers so if we have some ancient peer who really should be gone, if it shows up, we'll catch that error because we are only looking for the latest.
</details>
See my comment above. It is safe to entirely `latestPeers` since this is the only line where it is being mutated during tests and it can't contain any ancient peers.

---

*[test/www/jxcore/lib/thaliMobileNativeTestUtils.js, line 103 at r3](https://reviewable.io:443/reviews/thaliproject/thali_cordovaplugin/1275#-KTnMA9ofX9uvwgEyBgq:-KTnMA9pyWdFy-zuLoUl:b-1nipx4) ([raw file](https://github.com/thaliproject/thali_cordovaplugin/blob/2a6a9b75451b6f75ba9f563128d397a0dc1cc083/test/www/jxcore/lib/thaliMobileNativeTestUtils.js#L103)):*
> ```JavaScript
> }
> 
> function multiConnectEmitter () {
> ```

well, now this function is never called and native callbacks are never registered. I suggest to leave only `Mobile('...').registerToNative` calls and remove function itself and `inherits` below.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/thaliproject/thali_cordovaplugin/1275#-:-KTnMsaUGFO6EpNRe9aI:bly2pwy)*
<!-- Sent from Reviewable.io -->

ThaliTester commented:
PR is added to the queue for testing as 1. task. (af75736)

ThaliTester commented:
Test (Fail) 88496963 build is completed (af75736)

See https://github.com/ThaliTester/TestResults/tree/88496963af75736_iOS_enabled_mock_and_low_level_tests_yaronyg/ for the logs

ThaliTester commented:
PR is added to the queue for testing as 1. task. (f60e9c5)

ThaliTester commented:
Test (Fail) 88496963 build is completed (f60e9c5)

See https://github.com/ThaliTester/TestResults/tree/88496963f60e9c5_iOS_enabled_mock_and_low_level_tests_yaronyg/ for the logs

ThaliTester commented:
PR is added to the queue for testing as 1. task. (c6d9a1a)

ThaliTester commented:
Test (Fail) 88496963 build is completed (c6d9a1a)

See https://github.com/ThaliTester/TestResults/tree/88496963c6d9a1a_iOS_enabled_mock_and_low_level_tests_yaronyg/ for the logs

ThaliTester commented:
PR is added to the queue for testing as 1. task. (b6c922d)

ThaliTester commented:
Test (Success) 88496963 build is completed (b6c922d)

See https://github.com/ThaliTester/TestResults/tree/88496963b6c922d_iOS_enabled_mock_and_low_level_tests_yaronyg/ for the logs

ThaliTester commented:
Test 88496963b6c922d(b6c922d) has failed

See https://github.com/ThaliTester/TestResults/tree/88496963b6c922d_iOS_enabled_mock_and_low_level_tests_yaronyg/ for the fail logs

ThaliTester commented:
PR is added to the queue for testing as 1. task. (8ec0a5b)

ThaliTester commented:
Test (Success) 88496963 build is completed (8ec0a5b)

See https://github.com/ThaliTester/TestResults/tree/884969638ec0a5b_iOS_enabled_mock_and_low_level_tests_yaronyg/ for the logs

ThaliTester commented:
Test 884969638ec0a5b(8ec0a5b) has failed

See https://github.com/ThaliTester/TestResults/tree/884969638ec0a5b_iOS_enabled_mock_and_low_level_tests_yaronyg/ for the fail logs

ThaliTester commented:
PR is added to the queue for testing as 1. task. (9726768)

ThaliTester commented:
Test (Success) 88496963 build is completed (9726768)

See https://github.com/ThaliTester/TestResults/tree/884969639726768_iOS_enabled_mock_and_low_level_tests_yaronyg/ for the logs

ThaliTester commented:
Test 884969639726768(9726768) has failed

See https://github.com/ThaliTester/TestResults/tree/884969639726768_iOS_enabled_mock_and_low_level_tests_yaronyg/ for the fail logs

ThaliTester commented:
PR is added to the queue for testing as 1. task. (3e19cac)

ThaliTester commented:
Test (Fail) 88496963 build is completed (3e19cac)

See https://github.com/ThaliTester/TestResults/tree/884969633e19cac_iOS_enabled_mock_and_low_level_tests_yaronyg/ for the logs

ThaliTester commented:
PR is added to the queue for testing as 1. task. (4f55f99)

yaronyg commented:
Review status: 3 of 20 files reviewed at latest revision, 3 unresolved discussions.

---

*[test/www/jxcore/bv_tests/testThaliMobileNativeDiscoveryCoordinated.js, line 127 at r2](https://reviewable.io:443/reviews/thaliproject/thali_cordovaplugin/1275#-KTiTvODfrfUXOJcaXMh:-KTp3mQKqouVFhPuN1Cn:b-ishuha) ([raw file](https://github.com/thaliproject/thali_cordovaplugin/blob/915c2f2e94c4a4ca06906786409c9947476c9976/test/www/jxcore/bv_tests/testThaliMobileNativeDiscoveryCoordinated.js#L127)):*
<details><summary><i>Previously, chapko (Eugene Chapko) wrote…</i></summary>
> I'm still not agree. The following will be the same as an old code:
> ```javascript
> var obj = {};
> var currentPeers = obj;
> latestPeers = obj;
> ```
> In your case you create 2 separate independent object. Initially they are both empty and look the same when serialized, but they do not affect each other anymore.
> 
> But it doesn't matter anymore, you have fixed failing test below by explicitly populating `latestPeers`.
</details>
You are absolutely right. The objects are different. But why would it have mattered? That wasn't the point of the initialization. The goal was always just to empty out the two variables. The fact that they used a syntax that happened to set both objects to the identical empty object was just a side effect.

---

*[test/www/jxcore/bv_tests/testThaliMobileNativeDiscoveryCoordinated.js, line 212 at r3](https://reviewable.io:443/reviews/thaliproject/thali_cordovaplugin/1275#-KTl4ELaQENAoh1j-2TS:-KTp4zIP6C3EZS3shvcD:b-yhtlcn) ([raw file](https://github.com/thaliproject/thali_cordovaplugin/blob/2a6a9b75451b6f75ba9f563128d397a0dc1cc083/test/www/jxcore/bv_tests/testThaliMobileNativeDiscoveryCoordinated.js#L212)):*
<details><summary><i>Previously, chapko (Eugene Chapko) wrote…</i></summary>
> See my comment above. It is safe to entirely `latestPeers` since this is the only line where it is being mutated during tests and it can't contain any ancient peers.
</details>
Eeek! You actually pointed out that my code doesn't do what I had wanted it to do. My goal was that latestPeers would only contain peers that hadn't been previously advertised. But that isn't what I'm actually checking for. I've updated the code and the variable name to make clearer what the goal is.

---

*[test/www/jxcore/lib/thaliMobileNativeTestUtils.js, line 120 at r2](https://reviewable.io:443/reviews/thaliproject/thali_cordovaplugin/1275#-KTivd3bqOfHpDsI0oaW:-KU4fQKQjMHmd3JctOLj:b-xcy02b) ([raw file](https://github.com/thaliproject/thali_cordovaplugin/blob/915c2f2e94c4a4ca06906786409c9947476c9976/test/www/jxcore/lib/thaliMobileNativeTestUtils.js#L120)):*
<details><summary><i>Previously, yaronyg (Yaron Y Goland) wrote…</i></summary>
> Fixed
</details>
This is not workable because multiConnectEmitter is more than just an event emitter, it has additional logic.

---

*[test/www/jxcore/lib/thaliMobileNativeTestUtils.js, line 103 at r3](https://reviewable.io:443/reviews/thaliproject/thali_cordovaplugin/1275#-KTnMA9ofX9uvwgEyBgq:-KTp7-C-Sx2sDJwyX5MP:b-90x58u) ([raw file](https://github.com/thaliproject/thali_cordovaplugin/blob/2a6a9b75451b6f75ba9f563128d397a0dc1cc083/test/www/jxcore/lib/thaliMobileNativeTestUtils.js#L103)):*
<details><summary><i>Previously, chapko (Eugene Chapko) wrote…</i></summary>
> well, now this function is never called and native callbacks are never registered. I suggest to leave only `Mobile('...').registerToNative` calls and remove function itself and `inherits` below.
</details>
Yes this is broken but what is more broken is that I ran only the Android tests and not the iOS tests as well. I need to fix #915. I fixed the emitter code in a way that follows the usual node conventions. I also opened #1350 because even when tests fail the runCoordinatedTests.js script doesn't return an error.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/thaliproject/thali_cordovaplugin/1275)*
<!-- Sent from Reviewable.io -->

yaronyg commented:
@chapko Please see my comments.

ThaliTester commented:
PR is added to the queue for testing as 2. task. (b27c91e)

ThaliTester commented:
Test (Success) 88496963 build is completed (4f55f99)

See https://github.com/ThaliTester/TestResults/tree/884969634f55f99_iOS_enabled_mock_and_low_level_tests_yaronyg/ for the logs

ThaliTester commented:
Test 884969634f55f99(4f55f99) has failed

See https://github.com/ThaliTester/TestResults/tree/884969634f55f99_iOS_enabled_mock_and_low_level_tests_yaronyg/ for the fail logs

ThaliTester commented:
Test (Success) 88496963 build is completed (b27c91e)

See https://github.com/ThaliTester/TestResults/tree/88496963b27c91e_iOS_enabled_mock_and_low_level_tests_yaronyg/ for the logs

ThaliTester commented:
Test 88496963b27c91e(b27c91e) has failed

See https://github.com/ThaliTester/TestResults/tree/88496963b27c91e_iOS_enabled_mock_and_low_level_tests_yaronyg/ for the fail logs

chapko commented:
Reviewed 1 of 6 files at r4, 1 of 3 files at r5, 1 of 4 files at r6, 5 of 9 files at r7, 4 of 6 files at r8, 2 of 3 files at r9, 3 of 4 files at r10, 1 of 1 files at r11.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[test/TestServer/utils/asserts.js, line 82 at r11](https://reviewable.io:443/reviews/thaliproject/thali_cordovaplugin/1275#-KUMKvXvdzokS2yBmLFp:-KUMKvXvdzokS2yBmLFq:b-m87fc0) ([raw file](https://github.com/thaliproject/thali_cordovaplugin/blob/b27c91e0fae222fa41b866d918dd0f64d71f5245/test/TestServer/utils/asserts.js#L82)):*
> ```JavaScript
>     isObject(value),
>     format('object expected, received: \'%s\'', value) +
>       message ? message : ''
> ```

NB: Why mix string concatenation with `format`? You can just add `message` as an argument to the format.

---

*[test/www/jxcore/bv_tests/testThaliMobileNativeDiscoveryCoordinated.js, line 127 at r2](https://reviewable.io:443/reviews/thaliproject/thali_cordovaplugin/1275#-KTiTvODfrfUXOJcaXMh:-KUMMIdXwhBlI7sffJjD:b-ofywyx) ([raw file](https://github.com/thaliproject/thali_cordovaplugin/blob/915c2f2e94c4a4ca06906786409c9947476c9976/test/www/jxcore/bv_tests/testThaliMobileNativeDiscoveryCoordinated.js#L127)):*
When I reviewed it first time (915c2f2) there was only one place where `latestPeers` variable was actually used:
```javascript
var samePeers = peersReceived.every(function (idAndGeneration) {
  return latestPeers[idAndGeneration];
});
t.ok(samePeers, 'We have received peers from the latest peer discovery');
```
But `latestPeers` became independent object and was never updated directly in tests, so the `t.ok(samePeers, ...)` above would always fail.

> The goal was always just to empty out the two variables.

According to the `samePeers` check there should have been a link between `latestPeers` and `currentPeers`. It is not just a side effect.

Right now it doesn't matter, because you made it clearer and explicit. So this is OK.

---

*[test/www/jxcore/bv_tests/testThaliMobileNativeDiscoveryCoordinated.js, line 212 at r3](https://reviewable.io:443/reviews/thaliproject/thali_cordovaplugin/1275#-KTl4ELaQENAoh1j-2TS:-KUMspCYWHdejLvD3Rev:b32w0x) ([raw file](https://github.com/thaliproject/thali_cordovaplugin/blob/2a6a9b75451b6f75ba9f563128d397a0dc1cc083/test/www/jxcore/bv_tests/testThaliMobileNativeDiscoveryCoordinated.js#L212)):*
<details><summary><i>Previously, yaronyg (Yaron Y Goland) wrote…</i></summary>
> Eeek! You actually pointed out that my code doesn't do what I had wanted it to do. My goal was that latestPeers would only contain peers that hadn't been previously advertised. But that isn't what I'm actually checking for. I've updated the code and the variable name to make clearer what the goal is.
</details>
OK

---

*[test/www/jxcore/bv_tests/testThaliMobileNativeWrapper.js, line 356 at r11](https://reviewable.io:443/reviews/thaliproject/thali_cordovaplugin/1275#-KUN1fqDexvKMbz4V40h:-KUN1fqDexvKMbz4V40i:bvyh0el) ([raw file](https://github.com/thaliproject/thali_cordovaplugin/blob/b27c91e0fae222fa41b866d918dd0f64d71f5245/test/www/jxcore/bv_tests/testThaliMobileNativeWrapper.js#L356)):*
> ```JavaScript
>   })
>   .catch(function (error) {
>     if (platform._isRealIOS) {
> ```

Why is this changed to real check while 11 lines above we have usual check (for android)?

---

*[test/www/jxcore/lib/thaliMobileNativeTestUtils.js, line 120 at r2](https://reviewable.io:443/reviews/thaliproject/thali_cordovaplugin/1275#-KTivd3bqOfHpDsI0oaW:-KUMwh8O_bh_ZI9n4641:b32w0x) ([raw file](https://github.com/thaliproject/thali_cordovaplugin/blob/915c2f2e94c4a4ca06906786409c9947476c9976/test/www/jxcore/lib/thaliMobileNativeTestUtils.js#L120)):*
<details><summary><i>Previously, yaronyg (Yaron Y Goland) wrote…</i></summary>
> This is not workable because multiConnectEmitter is more than just an event emitter, it has additional logic.
</details>
OK

---

*[thali/NextGeneration/thaliMobile.js, line 32 at r11](https://reviewable.io:443/reviews/thaliproject/thali_cordovaplugin/1275#-KUN3yAgzIxWyQoiCXWL:-KUN3yAgzIxWyQoiCXWM:bm3uv5z) ([raw file](https://github.com/thaliproject/thali_cordovaplugin/blob/b27c91e0fae222fa41b866d918dd0f64d71f5245/thali/NextGeneration/thaliMobile.js#L32)):*
> ```JavaScript
>     return failure;
>   });
> };
> ```

I don't get the point of this function. It just swallows errors and returns promise fulfilled with the failure value. Why don't we just use promise states? It just seems natural for this.

---

*[thali/NextGeneration/thaliMobile.js, line 195 at r11](https://reviewable.io:443/reviews/thaliproject/thali_cordovaplugin/1275#-KUN4lIjdNVVYXt5WS6M:-KUN4lIjdNVVYXt5WS6N:b2ttb1k) ([raw file](https://github.com/thaliproject/thali_cordovaplugin/blob/b27c91e0fae222fa41b866d918dd0f64d71f5245/thali/NextGeneration/thaliMobile.js#L195)):*
> ```JavaScript
>         }
>         return reject(result);
>       });
> ```

The most expected thing here would be
```
.then(resolve, reject);
```
But we are using `promiseResultSuccessOrFailure` to join all failed and success results just to split them again here. It doesn't make sense to me. Also if one of the lower level `start` method some day will be resolved with non null value, it will treated as a failure by thaliMobile.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/thaliproject/thali_cordovaplugin/1275)*
<!-- Sent from Reviewable.io -->

ThaliTester commented:
PR is added to the queue for testing as 1. task. (1c94703)

ThaliTester commented:
Test (Success) 88496963 build is completed (1c94703)

See https://github.com/ThaliTester/TestResults/tree/884969631c94703_iOS_enabled_mock_and_low_level_tests_yaronyg/ for the logs

ThaliTester commented:
Test 884969631c94703(1c94703) has failed

See https://github.com/ThaliTester/TestResults/tree/884969631c94703_iOS_enabled_mock_and_low_level_tests_yaronyg/ for the fail logs

ThaliTester commented:
PR is added to the queue for testing as 1. task. (ea5ebba)

yaronyg commented:
Review status: 18 of 22 files reviewed at latest revision, 6 unresolved discussions.

---

*[test/TestServer/utils/asserts.js, line 82 at r11](https://reviewable.io:443/reviews/thaliproject/thali_cordovaplugin/1275#-KUMKvXvdzokS2yBmLFp:-KUTTjrM8DTOMjStCyE6:b-qg5mrg) ([raw file](https://github.com/thaliproject/thali_cordovaplugin/blob/b27c91e0fae222fa41b866d918dd0f64d71f5245/test/TestServer/utils/asserts.js#L82)):*
<details><summary><i>Previously, chapko (Eugene Chapko) wrote…</i></summary>
> NB: Why mix string concatenation with `format`? You can just add `message` as an argument to the format.
</details>
Because I'm lazy.

Fixed.

---

*[test/www/jxcore/bv_tests/testThaliMobileNativeDiscoveryCoordinated.js, line 127 at r2](https://reviewable.io:443/reviews/thaliproject/thali_cordovaplugin/1275#-KTiTvODfrfUXOJcaXMh:-KUTTtvXC8t3LW7aSbgQ:b-896fix) ([raw file](https://github.com/thaliproject/thali_cordovaplugin/blob/915c2f2e94c4a4ca06906786409c9947476c9976/test/www/jxcore/bv_tests/testThaliMobileNativeDiscoveryCoordinated.js#L127)):*
<details><summary><i>Previously, chapko (Eugene Chapko) wrote…</i></summary>
> When I reviewed it first time (915c2f2) there was only one place where `latestPeers` variable was actually used:
> ```javascript
> var samePeers = peersReceived.every(function (idAndGeneration) {
>   return latestPeers[idAndGeneration];
> });
> t.ok(samePeers, 'We have received peers from the latest peer discovery');
> ```
> But `latestPeers` became independent object and was never updated directly in tests, so the `t.ok(samePeers, ...)` above would always fail.
> 
> > The goal was always just to empty out the two variables.
> 
> According to the `samePeers` check there should have been a link between `latestPeers` and `currentPeers`. It is not just a side effect.
> 
> Right now it doesn't matter, because you made it clearer and explicit. So this is OK.
</details>
Done.

---

*[test/www/jxcore/bv_tests/testThaliMobileNativeDiscoveryCoordinated.js, line 212 at r3](https://reviewable.io:443/reviews/thaliproject/thali_cordovaplugin/1275#-KTl4ELaQENAoh1j-2TS:-KUTTxSFJgaDxXpwEip-:b-896fix) ([raw file](https://github.com/thaliproject/thali_cordovaplugin/blob/2a6a9b75451b6f75ba9f563128d397a0dc1cc083/test/www/jxcore/bv_tests/testThaliMobileNativeDiscoveryCoordinated.js#L212)):*
<details><summary><i>Previously, chapko (Eugene Chapko) wrote…</i></summary>
> OK
</details>
Done.

---

*[test/www/jxcore/bv_tests/testThaliMobileNativeWrapper.js, line 356 at r11](https://reviewable.io:443/reviews/thaliproject/thali_cordovaplugin/1275#-KUN1fqDexvKMbz4V40h:-KUTU7CJFocjB3pf-i6j:b-ivmjzd) ([raw file](https://github.com/thaliproject/thali_cordovaplugin/blob/b27c91e0fae222fa41b866d918dd0f64d71f5245/test/www/jxcore/bv_tests/testThaliMobileNativeWrapper.js#L356)):*
<details><summary><i>Previously, chapko (Eugene Chapko) wrote…</i></summary>
> Why is this changed to real check while 11 lines above we have usual check (for android)?
</details>
Real Android - calling killConnections MUST return a 'Not Supported' error. Which is what we check for.

Fake Android using wifi mock - Same behavior as Real Android

Real iOS - It's supposed to return success even though it's not actually implemented. After all, killConnections was invented for iOS only.

iOS using wifi mock - The wifi mock still has Android's behavior so it's going to fail in the same way as a above. This is what leads us to the _isRealIOS check. If this is the real iOS then we shouldn't fail. But if it's really the wifi mock then failure is fine.

---

*[test/www/jxcore/lib/thaliMobileNativeTestUtils.js, line 120 at r2](https://reviewable.io:443/reviews/thaliproject/thali_cordovaplugin/1275#-KTivd3bqOfHpDsI0oaW:-KUTVJcbc5QDRoV-gVVA:b-896fix) ([raw file](https://github.com/thaliproject/thali_cordovaplugin/blob/915c2f2e94c4a4ca06906786409c9947476c9976/test/www/jxcore/lib/thaliMobileNativeTestUtils.js#L120)):*
<details><summary><i>Previously, chapko (Eugene Chapko) wrote…</i></summary>
> OK
</details>
Done.

---

*[thali/NextGeneration/thaliMobile.js, line 32 at r11](https://reviewable.io:443/reviews/thaliproject/thali_cordovaplugin/1275#-KUN3yAgzIxWyQoiCXWL:-KUTVY_hCP6i5-Ry_KGG:b-mutq97) ([raw file](https://github.com/thaliproject/thali_cordovaplugin/blob/b27c91e0fae222fa41b866d918dd0f64d71f5245/thali/NextGeneration/thaliMobile.js#L32)):*
<details><summary><i>Previously, chapko (Eugene Chapko) wrote…</i></summary>
> I don't get the point of this function. It just swallows errors and returns promise fulfilled with the failure value. Why don't we just use promise states? It just seems natural for this.
</details>
I should use this code as in interview question. 'Explain, dear interviewee, why this is crap code'.

The answer is - because it's too damn subtle!

See the new comment I added

---

*[thali/NextGeneration/thaliMobile.js, line 195 at r11](https://reviewable.io:443/reviews/thaliproject/thali_cordovaplugin/1275#-KUN4lIjdNVVYXt5WS6M:-KUTWm1En8D-2kRnphMU:b4v845c) ([raw file](https://github.com/thaliproject/thali_cordovaplugin/blob/b27c91e0fae222fa41b866d918dd0f64d71f5245/thali/NextGeneration/thaliMobile.js#L195)):*
<details><summary><i>Previously, chapko (Eugene Chapko) wrote…</i></summary>
> The most expected thing here would be
> ```
> .then(resolve, reject);
> ```
> But we are using `promiseResultSuccessOrFailure` to join all failed and success results just to split them again here. It doesn't make sense to me. Also if one of the lower level `start` method some day will be resolved with non null value, it will treated as a failure by thaliMobile.
</details>
This has to do with a flaw in promise.All. If any promise fails then it returns immediately with just the failed promise's output. But for management purposes up the stack we really need to know the outcome of both promises. But there really isn't a Promise API I'm aware of that does this.

Things get more complicated because originally the design of the code was that we didn't fail, no matter what. But as I used the code it was clearly that this was a bad design that was really confusing when writing code to deal with thaliManager (who was just passing along what thaliMobile had to say). So I changed it.

So now we get both results no matter what. If they both succeed then we resolve() and if at least one fails then we reject(). I believe that reflects what most folks would expect.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/thaliproject/thali_cordovaplugin/1275)*
<!-- Sent from Reviewable.io -->

ThaliTester commented:
Test (Success) 88496963 build is completed (ea5ebba)

See https://github.com/ThaliTester/TestResults/tree/88496963ea5ebba_iOS_enabled_mock_and_low_level_tests_yaronyg/ for the logs

ThaliTester commented:
Test 88496963ea5ebba(ea5ebba) has failed

See https://github.com/ThaliTester/TestResults/tree/88496963ea5ebba_iOS_enabled_mock_and_low_level_tests_yaronyg/ for the fail logs
